Change hostname on aws instances from default to sct node name

aleksbykov opened the pull request:
Trello: https://trello.com/c/z6UKwDo1

Aws instances get default hostname based on private ip.
Now we change hostname right after node created and started
to sct node name.

## PR pre-checks (self review)
<!--- PR should be created as Draft, when CI finished and relevant checkboxes selected, add reviewers and then click on "Ready for review" button.-->
<!--- Put an `x` in all the boxes that apply or create PR and then click on all relevant checkboxes: -->
- [ ] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [ ] I gave variables/functions meaningful self-explanatory names
- [ ] I didn't leave commented-out/debugging code
- [ ] I didn't copy-paste code
- [ ] I added the relevant `backport` labels
- [ ] New configuration option are added and documented (in `sdcm/sct_config.py`)
- [ ] I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)
- [ ] All new and existing unit tests passed (CI)
- [ ] I have updated the Readme/doc folder accordingly (if needed)

aleksbykov commented:
Run manually. No any critical issue were found. Job log:https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/longevity-abykov-staging/job/longevity-1tb-7days-2Disruptive-Nemesises/
But at the begginging while node is rebooted after hostname changes, several errors for backtrace_threading appeared.
Investigating and fixing

aleksbykov commented:
> Run manually. No any critical issue were found. Job log:https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/longevity-abykov-staging/job/longevity-1tb-7days-2Disruptive-Nemesises/
> But at the begginging while node is rebooted after hostname changes, several errors for backtrace_threading appeared.
> Investigating and fixing

Fixed

fruch commented on sdcm/cluster.py:
```diff
@@ -2325,6 +2325,17 @@ def create_swap_file(self, size=1024):
+        result = self.remoter.run('sudo hostnamectl set-hostname {}'.format(self.name), ignore_status=True)
+        if result.ok:
+            self.log.info('Hostname has been changed. Rebooting...')
+            self.hard_reboot()
```
do we really have to restart it ?

fruch replied:
wouldn't be enough to re-login/reconnect ?

fruch replied:
we are adding this for cosmetics reason, I'm not sure adding a reboot per node for this worth the time it would take.

aleksbykov replied:
in manual, it is written to reboot, so all running processes get new hostname.
Will investigate more.

fruch replied:
one more idea it to try it in the startup_script, before we initially connect to the node.

aleksbykov replied:
@fruch do we have something similar on gce?

fruch replied:
no, in gce it's automatically set the hostname by the instance name. don't know exactly how that happens.(google didn't told me :smile:)

aleksbykov replied:
@fruch , with start_script it is a bit difficult, because we create instance before we create ( and generate) Node name

aleksbykov replied:
@fruch just reconnecting doesn't help, because the database log continue to use old hostname:
Dec 10 11:22:58 ip-172-30-0-111.ec2.internal

fruch replied:
Exactly why I've suggested using the run script, todo it before configuration of syslogd

aleksbykov replied:
@fruch I think adding the node name to start_script will require to much refactor (from my point of view): Now all instances have same start_script, if put code which set the hostname into it, we will need start_script per instances. Also for that moment we create all instances per cluster type with one call ec2_service.create_instances, otherwise it will require creating one by one.

fruch replied:
maybe just a restart to syslogd would be enough ? rebooting an i3 mean scylla data is lost, and might raise all kind of errors in the process, we should be careful with those.

aleksbykov replied:
Restarting systemd-journald switch to new hostname:
```
-- Logs begin at Wed 2019-12-11 06:48:56 GMT. --
Dec 11 06:49:12 ip-172-30-0-142.ec2.internal systemd[1]: Starting Scylla AMI Setup...
Dec 11 06:49:26 abykov-test-abykov-db-node-e8b4df74-1 sudo[1864]:     root : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/bin/sed -i s/#MaxSessions \(.*\)$/MaxSessions 1000/ /etc/ssh/sshd_config
Dec 11 06:49:26 abykov-test-abykov-db-node-e8b4df74-1 sudo[1868]:     root : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/bin/systemctl restart sshd
Dec 11 06:49:26 abykov-test-abykov-db-node-e8b4df74-1 sudo[1874]:     root : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/bin/sed -i s/#MaxSessions \(.*\)$/MaxSessions 1000/ /etc/ssh/sshd_config
Dec 11 06:49:26 abykov-test-abykov-db-node-e8b4df74-1 sudo[1880]:     root : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/bin/systemctl restart sshd
Dec 11 06:49:26 abykov-test-abykov-db-node-e8b4df74-1 scylla_ami_setup[1547]: Waiting for cloud-init to finish...
Dec 11 06:49:26 abykov-test-abykov-db-node-e8b4df74-1 scylla_ami_setup[1547]: [INFO] Started with user data set to:
```

amoskong replied:
> Restarting systemd-journald switch to new hostname:
> 
> ```
> -- Logs begin at Wed 2019-12-11 06:48:56 GMT. --
> Dec 11 06:49:12 ip-172-30-0-142.ec2.internal systemd[1]: Starting Scylla AMI Setup...
> Dec 11 06:49:26 abykov-test-abykov-db-node-e8b4df74-1 sudo[1864]:     root : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/bin/sed -i s/#MaxSessions \(.*\)$/MaxSessions 1000/ /etc/ssh/sshd_config
> Dec 11 06:49:26 abykov-test-abykov-db-node-e8b4df74-1 sudo[1868]:     root : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/bin/systemctl restart sshd
> Dec 11 06:49:26 abykov-test-abykov-db-node-e8b4df74-1 sudo[1874]:     root : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/bin/sed -i s/#MaxSessions \(.*\)$/MaxSessions 1000/ /etc/ssh/sshd_config
> Dec 11 06:49:26 abykov-test-abykov-db-node-e8b4df74-1 sudo[1880]:     root : TTY=unknown ; PWD=/ ; USER=root ; COMMAND=/bin/systemctl restart sshd
> Dec 11 06:49:26 abykov-test-abykov-db-node-e8b4df74-1 scylla_ami_setup[1547]: Waiting for cloud-init to finish...
> Dec 11 06:49:26 abykov-test-abykov-db-node-e8b4df74-1 scylla_ami_setup[1547]: [INFO] Started with user data set to:
> ```

In my manual test, journalctl switched to new hostname without restarting systemd-journald.

aleksbykov replied:
I ran several times, and it not happend.

bentsi commented on sdcm/cluster_aws.py:
```diff
@@ -648,6 +648,7 @@ def destroy(self):
         self.log.info('Destroyed')
 
     def start_task_threads(self):
+        self.change_hostname()
```
this is not a task thread, should be during init of the node

aleksbykov replied:
@bentsi Create newc commit, implementing via function in common.py.

bentsi replied:
how did you understand that you need to move this to separate module? Please move it back to the Node class
I meant that `change_hostname` is not a task thread, it shouldn't be started in `start_task_threads` method... (in other words it is a wrong place to put the call to this method)

aleksbykov replied:
yes, it was clear. The idea was to run change_hostname as earlier as possible, to collect logs with new hostname and for AWSNode only. The start_task_thread was best place, when AWSNode and Basenode Constructor initiated all required variables (new name, remoter).

amoskong replied:
Can we change the hostname by startup script? that's early enough

aleksbykov replied:
`I think adding the node name to start_script will require to much refactor (from my point of view): Now all instances have same start_script, if put code which set the hostname into it, we will need start_script per instances. Also for that moment we create all instances per cluster type with one call ec2_service.create_instances, otherwise it will require creating one by one.`
reason described before.

bentsi commented:
most of the CI tests failed.

amoskong commented on sdcm/cluster_aws.py:
```diff
@@ -437,6 +437,7 @@ def __init__(self, ec2_instance, ec2_service, credentials, parent_cluster,  # py
                                       node_prefix=node_prefix,
                                       dc_idx=dc_idx)
         if not cluster.Setup.REUSE_CLUSTER:
+            self.change_hostname()
```
Can you also change hostname for GCE instances?

aleksbykov replied:
@bentsi @fruch Should we?

fruch replied:
I think it's happening automatically already in GCE

bentsi replied:
@amoskong it happens automatically in GCE

amoskong replied:
Thanks.

aleksbykov commented:
@ bentsi, provisioning test failed due to error:
`No db log from node [Node PR-provision-test-jenkins-db-node-1777f66c-1 [18.203.139.9 | 10.0.55.35] (seed: True)] `

aleksbykov commented:
> @bentsi , provisioning test failed due to error:
> `No db log from node [Node PR-provision-test-jenkins-db-node-1777f66c-1 [18.203.139.9 | 10.0.55.35] (seed: True)] `
@bentsi @fruch 
Fixed an issue. It is not clear why, but for first node link to message.log -> database.log was not created. Moved the change hostname to BaseNode, fix this issue. But in this case name will be changed on any cluster types (aws, aws-cloud, gce, possible baremetal).

fruch commented on sdcm/cluster_aws.py:
```diff
@@ -431,7 +431,6 @@ def __init__(self, ec2_instance, ec2_service, credentials, parent_cluster,  # py
                                       node_prefix=node_prefix,
                                       dc_idx=dc_idx)
         if not cluster.Setup.REUSE_CLUSTER:
-
```
Please use `git revert` on this file

bentsi commented on sdcm/cluster.py:
```diff
@@ -2325,6 +2326,16 @@ def create_swap_file(self, size=1024):
         if "sct_configured_swapfile" not in result.stdout:
             self.log.warning("Swap file is not used on loader node %s.\nError details: %s", self, result.stderr)
 
+    def change_hostname(self):
```
If you put it here it will run for any class that inherits from BaseNode. 
Ask yourself: do we need this method to run in GCE? Do we need this in Docker? The answer is no.
So better:
- make this method abstract (raise not implemented error) in BaseNode
- change the name to `set_hostname` (we *set*/configure the name of the host)
- for AWS use this implementation
- for GCE skip

aleksbykov replied:
Done

bentsi commented on sdcm/cluster.py:
```diff
@@ -2325,6 +2326,9 @@ def create_swap_file(self, size=1024):
             self.log.warning("Swap file is not used on loader node %s.\nError details: %s", self, result.stderr)
 
+    def set_hostname(self):
+        self.log.warning('Method is not implemented for %s' % self.__class__.__name__)
```
next time : `raise NotImplementedError('Method is not implemented for %s' % self.__class__.__name__)` since we will not know when to implement it when adding new backend
